File: Free SysLog/Support Code/DataGridViewComparer.vb
Public Class DataGridViewComparer
    Implements IComparer(Of MyDataGridViewRow)

    Private ReadOnly intColumnNumber As Integer
    Private ReadOnly soSortOrder As SortOrder

    Public Sub New(columnNumber As Integer, sortOrder As SortOrder)
        intColumnNumber = columnNumber
        soSortOrder = sortOrder
    End Sub

    Public Function Compare(row1 As MyDataGridViewRow, row2 As MyDataGridViewRow) As Integer Implements IComparer(Of MyDataGridViewRow).Compare
        ' Compare them.
        If intColumnNumber = 0 Then
            Dim date1 As Date = row1.DateObject
            Dim date2 As Date = row2.DateObject

            Return If(soSortOrder = SortOrder.Ascending, Date.Compare(date1, date2), Date.Compare(date2, date1))
        Else
            Dim strFirstString As String = If(row1.Cells.Count <= intColumnNumber, "", row1.Cells(intColumnNumber).Value?.ToString())
            Dim strSecondString As String = If(row2.Cells.Count <= intColumnNumber, "", row2.Cells(intColumnNumber).Value?.ToString())

            Return If(soSortOrder = SortOrder.Descending, String.Compare(strFirstString, strSecondString), String.Compare(strSecondString, strFirstString))
        End If

        Return 0
    End Function
End Class

File: Free SysLog/Support Code/DataGridViewComparer.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;

namespace Free_SysLog
{
    public class DataGridViewComparer : IComparer<MyDataGridViewRow>
    {

        private readonly int intColumnNumber;
        private readonly SortOrder soSortOrder;

        public DataGridViewComparer(int columnNumber, SortOrder sortOrder)
        {
            intColumnNumber = columnNumber;
            soSortOrder = sortOrder;
        }

        public int Compare(MyDataGridViewRow row1, MyDataGridViewRow row2)
        {
            // Compare them.
            if (intColumnNumber == 0)
            {
                var date1 = row1.DateObject;
                var date2 = row2.DateObject;

                return soSortOrder == SortOrder.Ascending ? DateTime.Compare(date1, date2) : DateTime.Compare(date2, date1);
            }
            else
            {
                string strFirstString = row1.Cells.Count <= intColumnNumber ? "" : (row1.Cells[intColumnNumber].Value?.ToString());
                string strSecondString = row2.Cells.Count <= intColumnNumber ? "" : (row2.Cells[intColumnNumber].Value?.ToString());

                return soSortOrder == SortOrder.Descending ? string.Compare(strFirstString, strSecondString) : string.Compare(strSecondString, strFirstString);
            }

            return 0;
        }
    }
}